pkg/ovsdb/transact.go

```diff
@@ -0,0 +1,711 @@
+	switch actual.(type) {
+	case int:
+		return condInteger(actual, fn, expected)
+	case float64:
+		return condReal(actual, fn, expected)
+	case bool:
+		return condBoolean(actual, fn, expected)
+	case string:
+		return condString(actual, fn, expected)
+	case libovsdb.OvsSet:
+		return condSet(actual, fn, expected)
+	case libovsdb.OvsMap:
+		return condMap(actual, fn, expected)
+	case libovsdb.UUID:
+		return condUUID(actual, fn, expected)
+	default:
+		return false, errors.New("type not supported")
+	}
+}
+
+func isCondTypeUUID(c interface{}) (bool, error) {
+	cond, ok := c.([]interface{})
+	if !ok {
+		return false, errors.New("malformed condition")
+	}
+	key, op, value, err := parseCond(cond)
+	if err != nil {
+		return false, err
+	}
+	if key != UUID {
+		return false, errors.New("failed to convert to string")
+	}
+	if op != FN_EQ {
+		return false, nil
+	}
+	_, err = uuid.Parse(value.(libovsdb.UUID).GoUUID) /* TODO: missing assertion */
+	return true, err
+}
+
+func doesWhereContainCondTypeUUID(where []interface{}) (bool, error) {
+	for _, c := range where {
+		cond, ok := c.([]interface{})
+		if !ok {
+			return false, errors.New("malformed condition")
+		}
+		ok, err := isCondTypeUUID(cond)
+		if err != nil {
+			return false, err
+		}
+		if !ok {
+			return false, nil
+		}
+	}
+	return false, nil
+
+}
+
+func isRowSelectedByWhere(row map[string]interface{}, where []interface{}) (bool, error) {
+	for _, c := range where {
+		cond, ok := c.([]interface{})
+		if !ok {
+			return false, errors.New("malformed condition")
+		}
+		ok, err := isRowSelectedByCond(row, cond)
+		if err != nil {
+			return false, err
+		}
+		if !ok {
+			return false, nil
+		}
+	}
+	return false, nil
+}
+
+func isRowSelectedByCond(row map[string]interface{}, where []interface{}) (bool, error) {
+	column, fn, expected, err := parseCond(where)
+	ok := false
+	if err != nil {
+		return ok, err
+	}
+	actual := row[column]
+	return condTypeSwitch(actual, fn, expected)
+}
+
+func reduceRowByColumns(row map[string]interface{}, columns []string) map[string]interface{} {
+	return row // FIXME: missing implementation
+}
+
+func oneMutateRow(row map[string]interface{}, mutation interface{}) map[string]interface{} {
+	// mutatoin := mutation.([]interface{})
+	return row // FIXME: missing implementation
+}
+
+func allMutateRow(row map[string]interface{}, mutations []interface{}) map[string]interface{} {
+	for _, mutate := range mutations {
+		row = oneMutateRow(row, mutate)
+	}
+	return row
+}
+
+func etcdGetTable(txn *Transaction, table string) {
+	prefix := makePrefix(txn.dbname, table)
+	etcdOp := clientv3.OpGet(prefix, clientv3.WithPrefix())
+	txn.etcdThen = append(txn.etcdThen, etcdOp)
+}
+
+func etcdPutRow(txn *Transaction, table, uuid string, row map[string]interface{}) error {
+	key := makeKey(txn.dbname, table, uuid)
+	setRowUUID(row, uuid)
+	val, err := makeValue(row)
+	if err != nil {
+		return err
+	}
+	etcdOp := clientv3.OpPut(key, val)
```

Comment:
you cannot do it outside of transaction

Reply:
as explained this is not the transaction only preparation for it